python/lsst/ip/diffim/detail/buildSingleKernelVisitor.cc

```diff
@@ -0,0 +1,98 @@
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <https://www.lsstcorp.org/LegalNotices/>.
+ */
+#include <memory>
+#include <string>
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+#include <Eigen/Core>
+#include "numpy/arrayobject.h"
+#include "ndarray/pybind11.h"
+
+#include "lsst/afw/math/Kernel.h"
+#include "lsst/afw/math/SpatialCell.h"
+#include "lsst/ip/diffim/BuildSingleKernelVisitor.h"
+#include "lsst/pex/policy/Policy.h"
+
+namespace py = pybind11;
+using namespace pybind11::literals;
+
+namespace lsst {
+namespace ip {
+namespace diffim {
+namespace detail {
+
+namespace {
+
+/**
+ * Wrap BuildSingleKernelVisitor and the factory function for one pixel type
+ *
+ * @tparam PixelT  Pixel type, e.g. `float`
+ * @param mod  pybind11 module
+ * @param[in] suffix  Class name suffix associated with PixeT, e.g. "F" for `float`
+ */
+template <typename PixelT>
+void declareBuildSingleKernelVisitor(py::module& mod, std::string const& suffix) {
+    py::class_<BuildSingleKernelVisitor<PixelT>, std::shared_ptr<BuildSingleKernelVisitor<PixelT>>,
+               afw::math::CandidateVisitor>
+        cls(mod, ("BuildSingleKernelVisitor" + suffix).c_str());
+
+    cls.def(py::init<afw::math::KernelList, pex::policy::Policy>(), "basisList"_a, "policy"_a);
+    cls.def(py::init<afw::math::KernelList, pex::policy::Policy, std::shared_ptr<Eigen::MatrixXd>>(),
+            "basisList"_a, "policy"_a, "hMat"_a);
+
+    cls.def("setSkipBuilt", &BuildSingleKernelVisitor<PixelT>::setSkipBuilt, "skip"_a);
+    cls.def("getNRejected", &BuildSingleKernelVisitor<PixelT>::getNRejected);
+    cls.def("getNProcessed", &BuildSingleKernelVisitor<PixelT>::getNProcessed);
+    cls.def("reset", &BuildSingleKernelVisitor<PixelT>::reset);
+    cls.def("processCandidate", &BuildSingleKernelVisitor<PixelT>::processCandidate, "candidate"_a);
+
+    mod.def("makeBuildSingleKernelVisitor", (std::shared_ptr<BuildSingleKernelVisitor<PixelT>>(*)(
+                                                afw::math::KernelList const&, pex::policy::Policy const&)) &
+                                                makeBuildSingleKernelVisitor<PixelT>,
+            "basisList"_a, "policy"_a);
+    mod.def("makeBuildSingleKernelVisitor",
+            (std::shared_ptr<BuildSingleKernelVisitor<PixelT>>(*)(
+                afw::math::KernelList const&, pex::policy::Policy const&, std::shared_ptr<Eigen::MatrixXd>)) &
```

Comment:
I'm a little confused. I thought that shared_ptr's to Eigen matrices was removed because it was causing problems. Is it only a problem when a shared_ptr to an Eigen::Matrix is returned?